### PR TITLE
[configoptional] Make Unmarshal into None[T] match unmarshal into (*T)(nil)

### PR DESCRIPTION
**Pull request by mx-psi**
#### Description

Makes it so that unmarshaling into a `None[T]` is equivalent to unmarshaling into a `(*T)(nil)`.
The goal is to be able to do #13109 without introducing any changes to the behavior.

Needs #13161.

#### Testing

Add a unit test to compare the behavior of unmarshaling into `(*T)(nil)` with that of unmarshaling into `None[T]`

#### Documentation

Amend docstrings

### CODE REVIEW

**Comment by mx-psi** [comment]
I will add a changelog if this does not get merged before today's release

File: config/configoptional/optional.go [hunk 1]
```diff
@@ -24,6 +24,10 @@ type Optional[T any] struct {
 
 	// hasValue indicates if the Optional has a value.
 	hasValue bool
+
+	// notNone is used to indicate that the Optional is not None.
+	// This is used to differentiate between None and Default.
+	notNone bool
```
Comment (evan-bradley) [inline config/configoptional/optional.go:30]: Not something we have to decide here, and it's minor since it's largely internal, but how would you feel about calling this variable `hasValue` and renaming `hasValue` to `hasSome`?

1. Makes it so that `hasValue` means there's a default/explicitly-set value associated with an Optional object, through `Default`, `Some`, or unmarshaling.
2. Makes it so `hasSome` means the Optional object has a value ready for use, through `Some` or unmarshaling.
3. The `Default` function will no longer set a value while setting `hasValue: false`.
4. Gets rid of the negative in `notNone` while keeping our preferred default as the zero value for the field.

My hope is that this would simplify, if only a bit, how we describe the state an Optional object is in. The downside is the name of the `HasValue` function wouldn't make as much sense anymore, and changing that would be breaking if we do it later, so we'd need to agree this is enough of an improvement to justify that. Also, "some" isn't necessarily a common concept, so perhaps working with "value" and "none" are better.

File: config/configoptional/optional.go [hunk 2]
```diff
@@ -131,13 +133,24 @@ var _ confmap.Unmarshaler = (*Optional[any])(nil)
 
 // Unmarshal the configuration into the Optional value.
 //
+// The behavior of this method depends on the state of the Optional:
+//   - None[T]: does nothing if the configuration is nil, otherwise it unmarshals into the zero value of T.
+//   - Some[T](val): equivalent to unmarshaling into a field of type T with value val.
+//   - Default[T](val), equivalent to unmarshaling into a field of type T with value val, but using val if the configuration is nil.
```
Comment (evan-bradley) [inline config/configoptional/optional.go]:
```suggestion
//   - Default[T](val), equivalent to unmarshaling into a field of type T with base value val,
//     using val without overrides from the configuration if the configuration is nil.
```
I think we should refine the wording on this point a little. I think removing "but" in the original wording would be sufficient, but I came up with a suggestion to make it a little more explicit.

File: config/configoptional/optional.go (hunk 1 above)
Comment (jmacd) [reply config/configoptional/optional.go:30]: Makes sense to me.

**Review by jmacd — APPROVED** [review]

File: config/configoptional/optional.go (hunk 1 above)
Comment (mx-psi) [reply config/configoptional/optional.go:30]: I think it would be a bit awkward to have a `HasValue` method that returns something other than the internal `hasValue`. I would rename `HasValue` to `HasSome` (or maybe `IsSome` which is [the wording used by Rust](https://doc.rust-lang.org/std/option/enum.Option.html#method.is_some)?), but given the change in public API, I want to do this in a separate PR.

I filed #13181 and assigned it to me, will work on it after merging this.

Comment (jade-guiton-dd) [reply config/configoptional/optional.go:30]: If we need an explicit flag for all three states, what about using a None/Default/Some enum? Something like `type OptionalStatus int`.

The only reason a single bool used to work is that we didn't need to differentiate None from Default(zero value), but clearly that's no longer viable.

Comment (evan-bradley) [reply config/configoptional/optional.go:30]: Great idea, I think an enum would actually be perfect if we can make that work. Off the top of my head, I don't see any reason why it couldn't.
